Bring in E3SM cime changes for past 3 months

jgfouca opened the pull request:
Major change list:

**Python**
- Minor improvements to bless_test_results.py
- Add mach-specific throughput checking to test machines
- Some tools were not loading environment before running which was causing problems on some platforms
- Improvements in robustness, error msgs of check_input_data
- configure.py new FakeCase class
- add FakeCase to fortran_unit_testing/run_tests.py
- system_tests_common.py  change to tput test
- env_batch.py add new CHARGE_ACCOUNT
- env_mach_specific.py - Allow config_machines.xml to have any case variable as an attribute,
- pes.py - also output threads, root
- generic_xml.py - remove minidom
- build.py - add USE_MOAB option
- case.py - use new CHARGE_ACCOUNT.  new SMP_PRESENT
- provenancy.py - update E3SM provenance.
- test_scheduler.py - tput tolerance.  new "R" modifier for single column tests.
- utils.py new get_charge_account method
- update_acme_tests.py - more tests and test suites.

**schema changes**
- config_compilers_v2.xsd:  Add MOAB
- config_machines.xsd:  Add CHARGE_ACCOUNT for charging to a different name then PROJECT used for directory paths
- config_machines_template.xml:  Also add CHARGE_ACCOUNT
- env_mach_specific.xsd: Change debug to upperBoolean

**Tools**
- new acme_cime_split to help with subtree splitting
- new tests for the run_acme script in scripts/test/run_acme_test

**xml changes**
- Change "debug" to "DEBUG" in all config_machines.xml

** data models **
- datm/cime_config/namelist_definition_datm.xml new 6hourly forcing data set

** mct driver **
- config_component.xml  new SMP_PRESENT, USE_MOAB, TEST_TPUT_TOLERANCE, CHARGE_ACCOUNT
- cime_comp_mod.F90  new option to shr_mem_getusage to also print.
- seq_diag_mct.F90  fix an E3SM diag

**timing**
- fix for BGQ

shr_reprosum_mod.F90 - optionally check for nans.

Rename some man3 pages in PIO1

Many E3SM config changes.


Test suite: scripts_regression_tests
Test baseline: 
Test namelist changes: 
Test status: bit for bit

Fixes ESMCI/cime#1332
Fixes ESMCI/cime#1626

User interface changes?: 

Update gh-pages html  N

Code review:

jgfouca commented:
Tests still running, please review in the meantime.

jgfouca commented on config/acme/machines/config_batch.xml:
```diff
@@ -35,27 +35,27 @@
+       <arg flag="--cwd" name="CASEROOT"/>
+       <arg flag="-A" name="PROJECT"/>
+       <arg flag="-t" name="JOB_WALLCLOCK_TIME"/>
+       <arg flag="-n" name=" ($TOTALPES + $PES_PER_NODE - 1)/$PES_PER_NODE"/>
```
Did PES_PER_NODE get removed?

jedwards4b replied:
It was renamed MAX_MPITASKS_PER_NODE

jgfouca commented on scripts/Testing/Testcases/ICP_script:
```diff
@@ -0,0 +1,108 @@
+
+cat >! user_nl_cice.mhT <<EOF
+maskhalo_bound = .true.
+maskhalo_dyn = .true.
```
I think Testing/Testcases/* can be removed.

jgfouca commented on scripts/Testing/testreporter.pl:
```diff
@@ -0,0 +1,678 @@
+#-------------------------------------------------------------------------------
+# testreporter.pl
+# Perl script that watches the CIME tests as they progress, and sends the reports to
+# the testdb application at csegweb.cgd.ucar.edu
```
Can this be removed?

jgfouca commented on scripts/lib/CIME/SystemTests/pet.py:
```diff
@@ -33,6 +33,10 @@ def _case_one_setup(self):
         # Need to redo case_setup because we may have changed the number of threads
         case_setup(self._case, reset=True)
 
+        # Hack until CIME queue selection is more dynamic
```
This can be removed.

jgfouca commented on scripts/lib/CIME/XML/env_mach_specific.py:
```diff
@@ -54,44 +54,43 @@ def populate(self, machobj):
         resource_nodes = self.get_nodes("resource_limits")
         if resource_nodes is not None:
-            nodes = self._compute_resource_actions(resource_nodes, compiler, debug, mpilib)
+            nodes = self._compute_resource_actions(resource_nodes, case)
```
Changes to this file are worth a close review.

jedwards4b replied:
This is what allows config_machines.xml to have any case variable as an attribute, looks good.

jgfouca commented on scripts/lib/CIME/case_st_archive.py:
```diff
@@ -20,45 +20,143 @@ def _get_archive_file_fn(copy_only):
-
-    If there are no cpl restart files, this will return []
+    Returns the datetime objects specifying the times of each file
+    Note we are assuming that the coupler restart files exist and are consistent with other component datenames
```
@mfdeakin-sandia made a lot of changes to this file.

jgfouca replied:
@mfdeakin-sandia please help

mfdeakin-sandia replied:
I'll look into it - are there tests for the multi driver code, so I can make certain I put them back correctly?

jedwards4b replied:
yes, MCC in cime_developer is one.

jgfouca commented on scripts/lib/CIME/test_scheduler.py:
```diff
@@ -526,8 +528,27 @@ def _xml_phase(self, test):
-                    logger.debug (" STOP_OPTION set to {}".format(stop_option[opt]))
-                    logger.debug (" STOP_N      set to {}".format(opti))
+
+                    logger.debug (" STOP_OPTION set to %s" %stop_option[opt])
```
These need to be converted back to python3-style.

jedwards4b replied:
Once you have completed testing with python2 and resolved the concerns raised in review I will run testing with python3.

jgfouca commented on src/drivers/mct/bld/testdir/env_build.xml:
```diff
@@ -0,0 +1,71 @@
+    urrently, these variables are not cached.
+    Note1: users SHOULD NOT modify BUILD_COMPETE below
+    this is done automatically by the scripts.
+    </header>
```
Suspicious: seeing many new files.

rljacob replied:
This entire directory "src/drivers/mct/bld" is not present in ESMCI cime.

jgfouca commented on src/externals/pio1/tests/testpio/perl5lib/XML/man3/XML_Lite.3:
```diff
@@ -1,213 +0,0 @@
-.\" ======================================================================
-.de Sh \" Subsection heading
-.br
-.if t .Sp
```
@jayeshkrishna should these have been removed?

jgfouca commented on src/externals/pio2/src/clib/ncparser.pl:
```diff
@@ -0,0 +1,362 @@
+
+my $netcdf_inc = "$netcdf_incdir/netcdf.h";
+my $pnetcdf_inc = "$pnetcdf_incdir/pnetcdf.h";
+
```
@jayeshkrishna

jedwards4b replied:
why is this here?

jgfouca commented on tests/unit/test_names.c:
```diff
@@ -0,0 +1,416 @@
+#define NUM_NETCDF_FLAVORS 4
+#define NDIM 3
+#define X_DIM_LEN 400
+#define Y_DIM_LEN 400
```
What is this new file?

jgfouca commented on tools/mapping/gen_mapping_files/runoff_to_ocn/runoff_map_tx01.nml:
```diff
@@ -0,0 +1,15 @@
+   gridtype     = 'obs'
+   file_roff    = '/fs/cgd/csm/inputdata/lnd/clm2/rtmdata/rdirc.05.061026'
+   file_ocn     = '/fs/cgd/csm/mapping/grids/tx0.1v2_070911.nc'
+   file_nn      = 'map_nn_tx01v2.nc '
```
More zombie files?

jedwards4b replied:
I think so

rljacob commented:
Regarding the src/drivers/mct/bld dir and its files, according to the ESMCI git history:
```
commit 193b7d0528a13fd70f2a880f3a2d3981e539b19e
Author: James Foucar <jgfouca@sandia.gov>
Date:   Fri Jun 9 12:42:17 2017 -0600

    Remove src/drivers/mct/bld
```

So how come when we subtree merged ESMCI to ACME, this directory was not deleted?

@jgfouca I think you should make a PR to ACME that removes this dir and redo the subtree split after its merged to master.

billsacks commented on src/drivers/mct/cime_config/config_component.xml:
```diff
@@ -1422,6 +1449,40 @@
     <desc>ocn2atm state mapping file decomp type</desc>
   </entry>
 
+  <entry id="OCN2GLC_FMAPNAME">
```
@jedwards4b You mean OCN to GLC? I have a feeling they're needed for ACME, even though they're not currently needed for CESM. Need to make sure this is done in a way that won't break CESM.

billsacks commented:
Wow, this is huge. A few thoughts:

(1) @jgfouca Before this gets merged to master, can you please add documentation summarizing the changes that are coming in here? This documentation should make it into the merge commit.

(2) I'm really concerned about regressions, partly based on @jedwards4b 's findings. I'm not sure what to do about that: I don't think I have the energy to carefully review 8000+ lines of changes, though I certainly appreciate the time others have spent reviewing these changes so far! @jedwards4b @mvertens @fischer-ncar - I feel we should run the full CESM prebeta test suite on these changes (with baseline comparisons) as at least one additional sanity test.

(3) At today's cime call, I'd like to brainstorm how we can avoid big bang merges like this in the future.

rljacob commented:
Regrading 3, this should be the last one of these giant merges. For various reasons, we were delayed in updating CIME on the E3SM side, which in turn delayed this update back.  We can increase the frequency after this which will make these much smaller.

billsacks commented:
Okay, thanks @rljacob

jgfouca commented:
@rljacob I was thinking the same think. ACME merges keep bringing deleted files back from the dead, so we can try deleting these files from ACME.

I'd rather not re-do the merge since I've already done significant conflict-resolving work. I'll just delete them in both places.

jgfouca commented:
I apologize to the CSEG folks for dropping these massive merges. I know it's annoying for all parties involved. As Rob said, this should be the last one.

billsacks commented:
Thanks, @jgfouca . And I realize that, however painful this is on our end, it must be much more so on your end....

jgfouca commented on config/config_tests.xml:
```diff
@@ -247,6 +247,7 @@ NODEFAIL          Tests restart upon detected node failure. Generates fake failu
     <STOP_OPTION>ndays</STOP_OPTION>
     <STOP_N>11</STOP_N>
     <DOUT_S>FALSE</DOUT_S>
+    <BUILD_THREADED>TRUE</BUILD_THREADED>
```
@jedwards4b isn't it better to do XML setting here than in the python?

jedwards4b replied:
ok

jgfouca commented on config/xml_schemas/config_machines.xsd:
```diff
@@ -71,6 +72,9 @@
 	<xs:element ref="PROJECT" minOccurs="0" maxOccurs="1"/>
+	<!-- CHARGE_ACCOUNT: The name of the account to charge for batch jobs
+	  can be overridden in environment or $HOME/.cime/config -->
+	<xs:element ref="SUBPROJECT" minOccurs="0" maxOccurs="1"/>
```
@jedwards4b good catch. I think they meant to put CHARGE_PROJECT here instead of SUBPROJECT. I have no idea what a SUBPROJECT is.

rljacob replied:
This is our solution to https://github.com/ESMCI/cime/issues/1332 which was "fixed in ACME" and now being brought in here.

jgfouca commented on config/xml_schemas/env_mach_specific.xsd:
```diff
@@ -5,7 +5,7 @@
 <xs:attribute name="lang" type="xs:NCName"/>
 <xs:attribute name="compiler" type="xs:string"/>
-<xs:attribute name="debug" type="xs:boolean"/>
+<xs:attribute name="DEBUG" type="upperBoolean"/>
```
@jedwards4b I don't know why this change was made (changing debug to DEBUG). Is that going to break you guys?

jgfouca replied:
@amametjanov , looks like you made this change. Care to chime in?

jedwards4b replied:
Isn't this tied to the changes in env_mach_specific?  It will require some changes in the config_machines.xml file on the cesm side, but I think that it is a good change.

jgfouca commented on scripts/Testing/Testcases/ICP_script:
```diff
@@ -0,0 +1,108 @@
+
```
Nope. zombies

jgfouca commented on scripts/lib/CIME/SystemTests/erp.py:
```diff
@@ -27,9 +27,6 @@ def __init__(self, case):
                              run_one_description = 'initial',
                              run_two_description = 'restart')
 
-    def _common_setup(self):
```
wasn't needed once the XML did this for us.

jgfouca commented on scripts/lib/CIME/SystemTests/pet.py:
```diff
@@ -49,3 +53,7 @@ def _case_two_setup(self):
         # Need to redo case_setup because we may have changed the number of threads
         case_setup(self._case, reset=True)
+
+        # Hack until CIME queue selection is more dynamic
```
don't sweat it, it's gone.

jgfouca commented on scripts/lib/CIME/XML/generic_xml.py:
```diff
@@ -79,9 +79,8 @@ def write(self, outfile=None):
-            doc = minidom.parseString(xmlstr)
             with open(outfile,'w') as xmlout:
-                doc.writexml(xmlout,addindent='  ')
+                xmlout.write(xmlstr)
```
@jedwards4b it's been a while, but I think I was having problems with minidom on some machines. I checked the new output and still looked pretty readable. We can revisit this if needed.

jedwards4b replied:
maybe a try, except construct?

jgfouca commented on scripts/lib/CIME/test_scheduler.py:
```diff
@@ -526,8 +528,27 @@ def _xml_phase(self, test):
+                        envtest.set_test_parameter("NTASKS_"+comp,"1")
+
+                    # Set latitude and longitude for the appropriate case
+                    # Below for ARM97, default SCM test case
```
@jedwards4b can you elaborate? I don't know how to do this.

jedwards4b commented on src/share/util/mct_mod.F90:
```diff
@@ -98,6 +98,7 @@ module mct_mod
    use m_GlobalSegMap       ,only: mct_gsMap_gstorage     => GlobalStorage
    use m_GlobalSegMap       ,only: mct_gsMap_ngseg        => ngseg
    use m_GlobalSegMap       ,only: mct_gsMap_nlseg        => nlseg
+   use m_GlobalSegMap       ,only: mct_gsMap_OP           => OrderedPoints
```
This line was removed because it was a redefinition of mct_gsMap_orderedPoints on line 106.  I believe you need to make a change in the land model from mct_geMap_OP to mct_gsMap_orderedPoints

jgfouca commented:
TODO:

1. Add description to PR
2. case_st_archive
3. @jayeshkrishna to look at PIO
4. CSEG run a pre-beta test
5. Bring-in st_archiver changes as separate PR?
6. mct_mod change
7. CESM config_machines.xml needs to be changed to DEBUG

jayeshkrishna commented on src/externals/pio2/src/clib/ncputgetparser.pl:
```diff
@@ -0,0 +1,352 @@
+#!/usr/bin/perl 
```
Same here, ncputgetparser.pl was also deleted (Sep 17, 2016) from pio2 master (https://github.com/PARALLELIO/ParallelIO.git) and is no longer needed

jayeshkrishna commented on src/externals/pio2/src/clib/pio_c_get_template.c:
```diff
@@ -0,0 +1,62 @@
+int PIO_function()
```
Same here, pio_c_get_template.c was also deleted (Sep 17, 2016) from pio2 master (https://github.com/PARALLELIO/ParallelIO.git) and is no longer needed

jayeshkrishna commented on src/externals/pio2/src/clib/pio_c_put_template.c:
```diff
@@ -0,0 +1,77 @@
+///
```
Same here, pio_c_put_template.c was also deleted (Sep 17, 2016) from pio2 master (https://github.com/PARALLELIO/ParallelIO.git) and is no longer needed

jayeshkrishna commented on src/externals/pio2/src/clib/pio_c_template.c:
```diff
@@ -0,0 +1,54 @@
+int PIO_function()
```
Same here, pio_c_template.c was also deleted (Sep 17, 2016) from pio2 master (https://github.com/PARALLELIO/ParallelIO.git) and is no longer needed

jgfouca commented:
@jedwards4b @mvertens I think this is ready to go. Please do your prebeta testing and whatever other testing you think you need to do.

rljacob commented on scripts/lib/CIME/test_scheduler.py:
```diff
@@ -526,8 +528,24 @@ def _xml_phase(self, test):
                     logger.debug (" STOP_OPTION set to {}".format(stop_option[opt]))
                     logger.debug (" STOP_N      set to {}".format(opti))
+
+                elif opt.startswith('R'):
```
@jedwards4b and think you said we don't need this "R" option but its doing more then setting mpi-serial.

jedwards4b replied:
I think that the R is fine, but the hardcoded A97 case at lines 545-547 are not.   On the cesm side these have been moved to components/cam/cime_config

rljacob replied:
Which file in cam/cime_config does this go?

jedwards4b replied:
config_compsets.xml

rljacob commented:
I did a major update to the description.

jedwards4b commented on config/cesm/machines/config_machines.xml:
```diff
@@ -307,16 +307,16 @@
 	<command name="load">mkl</command>
       </modules>
-      <modules compiler="intel" mpilib="!mpi-serial" debug="true">
+      <modules compiler="intel" mpilib="!mpi-serial" DEBUG="true">
```
I believe that the TRUE FALSE also needs to be upcased.

jedwards4b commented:
There was a problem in the schema file for env_mach_specific.xml - the attribute modifiers were defined for both environment_variables and for env - you can't have both (okay technically you can but it makes for a logic nightmare) I have removed the attributes from the env field -  this will probably require changes in the acme config_machines.xml file.

jgfouca commented:
@jedwards4b , we've already removed all the attribute modifiers for individual environs in ACME.

jedwards4b commented:
Good, no problem then.

jgfouca commented:
@jedwards4b I've addressed your feedback regarding the 'R' options.

@rljacob I've left a comment in test_scheduler that will hopefully remind us to make the needed changes once this gets back to ACME.

rljacob commented:
ok

jgfouca commented:
@jedwards4b OK, i will go ahead and merge then.
